sympy/ntheory/primetest.py

```diff
@@ -204,7 +207,9 @@ def isprime(n):
     sympy.ntheory.generate.prime : Return the nth prime
 
     """
-    n = int(n)
+    if isinstance(n, (float, Float)):
+        return False
+    n = as_int(n)
```

Comment:
Is as_int(n) correct here because it fails when n = ( 3 + ( sin(1)**2 + cos(1)**2 ) ). Should I revert it to int(n)

Reply:
I think so. This contrived example was to show that some expressions are ints...and in this case it is not an approximation, it actually is/would be an int. But I think the consensus is that this should only be taking int input. Anything else requires user-preconditioning. The `as_int` converts Integer to int which increases the performance of the ntheory routines.

Reply:
There will be two test cases fails in [test_query.py](https://github.com/sympy/sympy/blob/master/sympy/assumptions/tests/test_query.py#L2101) after taking int input only in isprime.

`assert ask(Q.prime(12345678901234567890 + (cos(1)**2 + sin(1)**2))) is True`
`assert ask(Q.prime(Add(S(2), S(2), evaluate=0))) is False`

Should I modify the input of these test cases or should I remove them ?

Reply:
I am guessing that the assumption system is trying to use isprime to answer the question first. Since isprime is not nuanced in return value -- it's yes or no -- and it gives a False for non-ints, the assumption system returns that value. So maybe what is needed is to raise a ValueError for non-int input so the user/caller knows that isprime is not going to answer such queries: if it's not an int, `isprime` doesn't answer.

Any other ideas?

Reply:
BTW, `factorint` raises a ValueError for non-int input.

Reply:
Is there any way to evaluate (cos(1)**2 + sin(1)**2)) symbolically first as it in fact is an int?

(Btw, there are no two binary numbers whose squares sums to exactly one, so doing it numerically will always give a number != 1 independent of the value put in, unless the floats happens to round to one.)

Reply:
I understand that the expression as it stands doesn't do it. I was more thinking of S.One or S(1) or simply put an arbitrary symbol there.

Reply:
I'm surprised that the test with the trigonometric identity fails. 

```
print((6 + (sin(1)**2 + cos(1)**2)).simplify())
7
```
I do not know how it is evaluated, but clearly there are mechanisms that keep integers as integers (note 7, not 7.0). That expression is currently prime by the way.

Reply:
simplify can pick up this identity but I don't think you want a low-level routine handling the simplification.
```python
>>> simplify(cos(x)**2+sin(x)**2)
1
```

Reply:
> I am guessing that the assumption system is trying to use isprime to answer the question first. Since isprime is not nuanced in return value -- it's yes or no -- and it gives a False for non-ints, the assumption system returns that value. So maybe what is needed is to raise a ValueError for non-int input so the user/caller knows that isprime is not going to answer such queries: if it's not an int, `isprime` doesn't answer.
> 
> Any other ideas?

I think that it is fine to raise Value Error for all non int inputs.


> BTW, `factorint` raises a ValueError for all non-int input.

`factorint` also uses `as_int`.
`factorint` accepts input value of `9.0` which will converted to `9` by `as_int`

Reply:
If `as_int` rejects Floats then the behavior will be consistent in both routines.